File: ML for Satyavrat/Djikstra.c
#include <stdio.h>
#include <limits.h>
#include <stdbool.h>

#define V 4

int minDistance(int dist[], bool visited[]) {
    int min = INT_MAX, min_index;

    for (int v = 0; v < V; v++) {
        if (!visited[v] && dist[v] <= min) {
            min = dist[v];
            min_index = v;
        }
    }
    return min_index;
}

void dijkstra(int graph[V][V], int src) {
    int dist[V];   
    bool visited[V];  

    for (int i = 0; i < V; i++) {
        dist[i] = INT_MAX;
        visited[i] = false;
    }

    dist[src] = 0;

    for (int count = 0; count < V - 1; count++) {
        int u = minDistance(dist, visited);

        visited[u] = true;

        for (int v = 0; v < V; v++) {
            if (!visited[v] && graph[u][v] && dist[u] != INT_MAX && dist[u] + graph[u][v] < dist[v]) {
                dist[v] = dist[u] + graph[u][v];
            }
        }
    }

    printf("Vertex \t Distance from Source\n");
    for (int i = 0; i < V; i++) {
        printf("%d \t %d\n", i, dist[i]);
    }
}

int main() {
    int graph[V][V] = {
        {0, 1, 4, 0},
        {1, 0, 2, 6},
        {4, 2, 0, 3},
        {0, 6, 3, 0}
    };

    int start_vertex = 0;
    dijkstra(graph, start_vertex);

    return 0;
}

File: ML for Satyavrat/Djikstra.py
# Define the number of vertices
V = 4

def min_distance(dist, visited):
    # Initialize minimum value
    min_value = 100000
    min_index = -1

    for v in range(V):
        if not visited[v] and dist[v] <= min_value:
            min_value = dist[v]
            min_index = v
    return min_index

def dijkstra(graph, src):
    # Use a high value for "infinity"
    INF = 100000

    # Initialize distance array and visited set
    dist = [INF] * V
    visited = [False] * V

    # Distance to the source itself is always 0
    dist[src] = 0

    for _ in range(V - 1):
        # Pick the minimum distance vertex from the set of vertices not yet processed
        u = min_distance(dist, visited)

        # Mark the vertex as processed
        visited[u] = True

        # Update distance values of adjacent vertices of the picked vertex
        for v in range(V):
            if (
                graph[u][v] > 0 and  # Edge exists
                not visited[v] and  # Not already visited
                dist[u] != INF and  # Source vertex distance is not INF
                dist[u] + graph[u][v] < dist[v]  # Update condition
            ):
                dist[v] = dist[u] + graph[u][v]

    # Print the results
    print("Vertex \t Distance from Source")
    for i in range(V):
        print(f"{i} \t {dist[i]}")

# Adjacency matrix representation of the graph
graph = [
    [0, 1, 4, 0],
    [1, 0, 2, 6],
    [4, 2, 0, 3],
    [0, 6, 3, 0]
]

# Run Dijkstra's algorithm
start_vertex = 0
dijkstra(graph, start_vertex)
